### PR TITLE
[11.0] [IMP] Increase speed in partner Kanban view.

### PR DESCRIPTION
**Pull request by feketemihai**
Description of the issue/feature this PR addresses:
Speeding up Kanban loading view of partners.

Current behavior before PR:
Kanban view is loading slowly for a lot of partners.

Desired behavior after PR is merged:
Increase the speed in loading. Difference between mapped on child_ids and search_count is more than 50% faster.

Even that a store in the db and to make function dependent will increase higher, probably to take it in consideration for v 12.0. There are a lot of count fields in res.partner, not only from crm module, but this one's are shoved in the Kanban view.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by feketemihai** [comment]
If you think it's fine, i can change all the count methods, from purchase, account, point of sale...

**Comment by astirpe** [comment]
@feketemihai thank you for this! 
I think your PR also fixes an [issue](https://travis-ci.org/onesteinbv/addons-onestein/jobs/357445666) we recently experienced while testing an onchange method.

And yes, I would be very happy happy if method `def _purchase_invoice_count()` is also fixed.

**Comment by pedrobaeza** [comment]
It's strange that a `len` operation over a recordset with `mapped` works better than a `read_group`, as this triggers a prefetching of all the fields of the subtable. When adding more fields to the base model, this can scale bad.

**Comment by feketemihai** [comment]
@pedrobaeza I will make tomorrow a comparition for all the count fields on res.partner, too see which one will gain more from this. For opportunities, it reduce under 1 ms while the initial calculation was from 2-3 ms.

**Comment by feketemihai** [comment]
[Performance_compare_demo_data.xlsx](https://github.com/odoo/odoo/files/1846579/Performance_compare_demo_data.xlsx)

**Comment by feketemihai** [comment]
@pedrobaeza @astirpe You can check performance in the table attached. Probably @rco-odoo can help us on this, to have a consistency in count functions, to have the best performance with kind of the same coding

**Comment by feketemihai** [comment]
Sorry, bad copy-paste on opportunity...
[Performance_compare_demo_data.xlsx](https://github.com/odoo/odoo/files/1846598/Performance_compare_demo_data.xlsx)

**Comment by pedrobaeza** [comment]
The problem then seems to be to not perform one `read_group` but several, isnt' it?

**Comment by feketemihai** [comment]
The problem is crm, sale was search_count  plus operator child_of, on purchases and POS we don't have a one2many defined in res.partner, so tried to make them consistent,even that len() over one2many will be faster...

File: addons/crm/models/res_partner.py [hunk 1]
```diff
@@ -3,7 +3,6 @@
 
 from odoo import api, fields, models
 
-
```
Comment (mreficent) [inline addons/crm/models/res_partner.py]: don't delete this line 😉

**Comment by tivisse** [comment]
@rco-odoo What do you think about this ? May I get your input ? 

Thanks !

File: addons/crm/models/res_partner.py [hunk 2]
```diff
@@ -16,9 +16,10 @@ class Partner(models.Model):
 
     @api.multi
     def _compute_opportunity_count(self):
+        partners_data = self.env['crm.lead'].read_group([('partner_id', 'in', self.ids), ('type', '=', 'opportunity')], ['partner_id'], ['partner_id'])
+        mapped_data = dict([(partner['partner_id'][0], partner['partner_id_count']) for partner in partners_data])
         for partner in self:
-            operator = 'child_of' if partner.is_company else '='  # the opportunity count should counts the opportunities of this company and all its contacts
-            partner.opportunity_count = self.env['crm.lead'].search_count([('partner_id', operator, partner.id), ('type', '=', 'opportunity')])
+            partner.opportunity_count = mapped_data.get(partner.id, 0) + sum(mapped_data.get(int(child), 0) for child in partner.mapped('child_ids'))
```
Comment (jem-odoo) [inline addons/crm/models/res_partner.py]: I don't think this code is correct. 
So here, we are summing the children count to the one from the parent. The parent (current partner in the loop) is obviously present in the `mapped_data` dict as the read_group domain takes ` self.ids`. But, the children might not be in `self.ids`, and so will not be present in the dict.
If you have tested and it was working, i guess the children was on your kanban view (for instance). I believe, if you tried with a domain excluding some children, the count of a parent partner will not be correct.

File: addons/sale/models/res_partner.py [hunk 3]
```diff
@@ -14,14 +14,7 @@ class ResPartner(models.Model):
...
-        mapped_data = dict([(m['partner_id'][0], m['partner_id_count']) for m in sale_data])
+        partners_data = self.env['sale.order'].read_group([('partner_id', 'in', self.ids)], ['partner_id'], ['partner_id'])
+        mapped_data = dict([(partner['partner_id'][0], partner['partner_id_count']) for partner in partners_data])
         for partner in self:
-            # let's obtain the partner id and all its child ids from the read up there
-            item = next(p for p in partner_child_ids if p['id'] == partner.id)
-            partner_ids = [partner.id] + item.get('child_ids')
-            # then we can sum for all the partner's child
-            partner.sale_order_count = sum(mapped_data.get(child, 0) for child in partner_ids)
+            partner.sale_order_count = mapped_data.get(partner.id, 0) + sum(mapped_data.get(int(child), 0) for child in partner.mapped('child_ids'))
```
Comment (jem-odoo) [inline addons/sale/models/res_partner.py]: same remarks as in CRM

**Pull request closed without merging**